### ISSUE TITLE
Add a function to compute easily best size for the space

### ISSUE THREAD

**Issue by pommedeterresautee** [wush978/FeatureHashing#52]
It seems there is an easy way to compute the best space size (not too big but still enough to limit collision). From [Vowpal Wabbit](https://github.com/JohnLangford/vowpal_wabbit/wiki/Feature-Hashing-and-Extraction):

> Ensuring the hash is large enough

> The hash table, by default, can hold 2^18 or 262144 entries. For many problems this is plenty but in some cases, more space is needed to avoid collisions. To count unique features after hashing, add the parameter --readable_model <fname> then use wc -l <fname>. Calculate the nearest power of two to the result and use this to set the -b parameter.

Counting unique feature per column in data.frame is fairly simple with unique() function and a call to apply(). Does it make sense to include such function?

**Comment by pommedeterresautee** [comment]
May be of interest:
http://stackoverflow.com/questions/6461209/how-to-round-up-to-the-nearest-10-or-100-or-x

**Comment by pommedeterresautee** [comment]
Or just copy paste the code from 
http://www.inside-r.org/packages/cran/pracma/docs/nextpow2

**Comment by wush978** [comment]
The FeatureHashing should be used in an situation where the global information, i.e. the number of unique feature, is not easily acquired. Therefore, I think this feature is ok but not necessary. I prefer to do this after receiving more requests from user.

**Comment by pommedeterresautee** [comment]
I get your point but I am thinking to one use case (not yet tested a lot) where everyone may be interested in feature hashing.

Converting dense df to sparse one hot encoded matrix takes a lot of time. When the df is big (> 4Gb for instance), there are some memory pick during the conversion, making mandatory to perform the conversion column per column. This process takes time. I need to do test first, but if feature hashing takes less time and less memory to convert the df to something which can be used by xgboost, I would advise in the xgboost documentation this kind of conversion instead of one hot encoding. I am realizing the model won't be exactly the same but it should be very similar (need test too).

Moreover, you may want to use hashing trick for feature reduction, and knowing the optimal space to avoid collision is required for that purpose too.

**Comment by wush978** [comment]
Your point is reasonable, so maybe we could arrange this to next patch.

**Comment by pommedeterresautee** [comment]
Clearly there is no urgency for this feature. Cleaning the API and improving the documentation are much more important.

**Comment by pommedeterresautee** [comment]
I just rewrote the documentation of this function, and now, I better understand why you wrote it s not that important to have such feature. On a functional point of view I fully agree, but I feel that with the explanation in the documentation, a user not familiar with hashing will have a better understanding of how hashing works. He may experiment easily how the collision rate decreases with the size and so on. That's why I still think it deserves to be part of the package. And as you wrote, it doesn't hurt.

Let me know if the documentation I wrote is correct (I understood things between the implementation and rewriting the documentation, so may be there are some other things I still didn't get).

Kind regards,
Michaël
